### PR TITLE
AB-297: Add docker support for development

### PR DESCRIPTION
**Pull request by ferbncode**
Task remaining: 

- [x] Updating documentation (`README.md`)
- [x] Removing some redundant dependencies and packages installed in some containers dataset_evaluator` and `hl_extractor` (and some code being copied).
- [x] Running tests
- [x] Removing vagrant related files

### CODE REVIEW

**Comment by ferbncode** [comment]
As stated in the ticket, I've added Dockerfiles for the `Webserver`, `Hl Calculator`, `Model Trainer` and the `database`.
+ After modifying the files `hl_extractor/profile.conf.in` and `config.py`, one can build all services using:
    `docker-compose -f docker/docker-compose.dev.yml build`
+ Intialize AB database using for the first time:
    `docker-compose -f docker/docker-compose.dev.yml run webserver python manage.py init_db --skip-create-db`
+ and all services can be run using:
    `docker-compose -f docker/docker-compose.dev.yml up`

**Comment by alastair** [comment]
Thanks! :+1:
I plan to review this the week of 18-22 September. If I haven't given any response by the summit please remind me!

**Comment by paramsingh** [comment]
Should we add a `develop.sh` as present in LB and MsB too?

**Comment by paramsingh** [comment]
@alastair, adding a Dockerfile.test would make testing with Jenkins easy.

File: docker/Dockerfile.gaia [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+COPY docs/requirements.txt /hl_extractor/docs/requirements.txt
+RUN pip install -r requirements.txt
+
+# Install Dockerize
+ENV DOCKERIZE_VERSION v0.2.0
+RUN wget https://github.com/jwilder/dockerize/releases/download/$DOCKERIZE_VERSION/dockerize-linux-amd64-$DOCKERIZE_VERSION.tar.gz \
+    && tar -C /usr/local/bin -xzvf dockerize-linux-amd64-$DOCKERIZE_VERSION.tar.gz
+
+COPY ./db /hl_extractor/db
+COPY ./default_config.py /hl_extractor/default_config.py
```
Comment (paramsingh) [inline docker/Dockerfile.gaia]: Why is this needed?

File: docker/docker-compose.dev.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    volumes:
+      - ../:/code
+      - ../data/app:/data
+    ports:
+      - "8080:8080"
+    depends_on:
+      - db
+      - acousticbrainz_cache
+
+  acousticbrainz_cache:
```
Comment (paramsingh) [inline docker/docker-compose.dev.yml]: We use redis for the container name in LB and MsB, renaming it for consistency might be a good idea.

File: docker/Dockerfile.gaia (hunk 1 above)
Comment (alastair) [reply docker/Dockerfile.gaia]: ah, I thought I cleaned up this, it was in the original Dockerfile but isn't needed

File: README.md [hunk 3]
```diff
@@ -50,10 +38,16 @@ Full installation instructions are available in [INSTALL.md](https://github.com/
...
+After you started development versions of containers with `docker-compose`, connect to the main container:
+
+`docker-compose -f docker/docker-compose.dev.yml run webserver run /bin/bash`
+
+Calling `gulp` on its own will build everything necessary
 to access the server in a web browser:
 
-    ./node_modules/.bin/gulp
+    root@<container_id>:/code# npm install
+    root@<container_id>:/code# ./node_modules/.bin/gulp
```
Comment (alastair) [inline README.md]: we need to work out a better way to do this - ideally we should install modules during Docker build, and then work out how to serve js and css during development and in production. I'm not sure how other MeB projects host static content, for my other projects I share a volume or host directory between my app container and nginx container and sync it during deploy

File: README.md [hunk 4]
```diff
@@ -6,40 +6,28 @@ The server components for the new AcousticBrainz project.
 Please report issues here: http://tickets.musicbrainz.org/browse/AB
 
 
-## Installation
+## Installation and Running
 
 ### Vagrant VM
```
Comment (alastair) [inline README.md]: Docker

File: README.md [hunk 5]
```diff
@@ -6,40 +6,28 @@ The server components for the new AcousticBrainz project.
...
 
+In order to load a psql session, use the following command:
 
-There are some shortcuts defined using fabric to perform commonly used
-commands:
-
- * `fab vpsql`: Load a psql session. Requires a local psql client
- * `fab vssh`: Connect to the VM more efficiently, saving the settings
-               so that you don't need to run vagrant each time you ssh.
+`docker-compose -f docker/docker-compose.dev.yml run webserver psql -U acousticbrainz -h db`
```
Comment (alastair) [inline README.md]: just thinking out loud about this - do we need the psql client in the web container? It's in the db container, so we can still have a variation of this line, but let's not install things that aren't needed in the web container

File: README.md [hunk 6]
```diff
@@ -77,37 +71,11 @@ to log in.
 
 Once you have logged in, you can make your user an admin, by running
 
-    python manage.py add_admin <your user>
+    docker-compose -f docker-compose.dev.yml run webserver python2 manage.py add_admin <your user>
```
Comment (alastair) [inline README.md]: in these run commands we should add the `--rm` flag to remove the container after it finishes

File: README.md [hunk 7]
```diff
@@ -117,16 +85,16 @@ Latest database dump is available at http://acousticbrainz.org/download. You
 need to download full database dump from this page and use it during database
 initialization:
 
-    $ python manage.py init_db path_to_the_archive
+    $ docker-compose -f docker/docker-compose.dev.yml run webserver python2 manage.py init_db path_to_the_archive
```
Comment (alastair) [inline README.md]: this will need volumes if the path isn't in the current directory and we're not mounting it as a volume

File: README.md (hunk 6 above)
Comment (alastair) [reply README.md]: ideally we should be using `-p` to set the project to `acousticbrainz` instead of the default `docker`. Perhaps adding in develop.sh would help with this?

Should we consider moving the main content of test.sh and develop.sh to brainzutils and have stubs in the projects? I'm not sure how to "install" it for development purposes, as brainzutils installs into the docker container

File: docker/Dockerfile.gaia (hunk 1 above)
Comment (alastair) [reply docker/Dockerfile.gaia]: in fact, this is the wrong file. I hope I didn't delete all of my changes

**Comment by alastair** [comment]
I'm happy with this PR now. I've done a clean checkout and run the commands directly from the readme to ensure that it results in a server that works. The hl extractor and dataset interface still works.
@paramsingh @ferbncode do you have anything further that you want to work on? My preference would be to merge this asap for SoC incoming students, and work with @zas separately to get it ready for production release.

File: docker/docker-compose.dev.yml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+  webserver:
+    build:
+      context: ..
+      dockerfile: ./docker/Dockerfile.dev
+    volumes:
+      - ../:/code
+      - ../data/app:/data
+    ports:
+      - "8080:8080"
```
Comment (paramsingh) [inline docker/docker-compose.dev.yml]: We use the 80 port in CB and LB.

Comment (alastair) [reply docker/docker-compose.dev.yml]: My preference for development is to use a non-standard port, in case 80 is already used. It's only a small difference so I'll leave it for now

**Comment by alastair** [comment]
Merged! Thanks @ferbncode and @paramsingh for your help!
